axiom/nr_attributes.c

```diff
@@ -162,10 +189,133 @@ static int nr_attribute_destination_modifier_compare(
+ *           Also note an empty include rule means include everything and
+ *           and to exclude nothing.
+ *
+ *           All other attributes besides log context attributes do NOT have
+ *           this exclusive behavior for the include rules.  This function only
+ *           considers rules of the NR_ATTRIBUTE_DESTINATION_LOG destination.
+ *
+ *           The algorithm examines all input rules for the
+ *           NR_ATTRIBUTE_DESTINATION_LOG destination, and will create a new
+ *           rule of "exclude=*" with "is_finalize_rule=true" if any input 
+ *           rules exist.  The exception is if there is an input rule of "*"
+ *           for destination NR_ATTRIBUTE_DESTINATION_LOG.  In this case no 
+ *           finalize rule is added.
+ *
+ *           The net resulting effect is the exclusion of any attributes not
+ *           contained in the set of input rules for
+ *           NR_ATTRIBUTE_DESTINATION_LOG. The exception for if an "input=*"
+ *           rule is necessary since this input rule excludes nothing, making
+ *           the finalize rule of "exclude=*" unnecessary.
+ *
+ */
+static void nr_attribute_config_finalize_log_destination(
+    nr_attribute_config_t* config) {
+  nr_attribute_destination_modifier_t* cur = NULL;
+  nr_attribute_destination_modifier_t* prev = NULL;
+  nr_attribute_destination_modifier_t* next = NULL;
+  bool add_finalize_rule = false;
+
+  if (NULL == config || NULL == config->modifier_list) {
+    /* Since there is no configuration, no work to do. */
+    return;
+  }
+
+  /* remove any existing rules with is_finalize_rule = true */
+  for (cur = config->modifier_list; cur;) {
+    next = cur->next;
+
+    /* currently only finalize rules being created are for the
+     * NR_ATTRIBUTE_DESTINATION_LOG destination but check to
+     * be thorough and in case other finalize rules are created
+     * in the future.
+     */
+    if (cur->is_finalize_rule
+        && (cur->include_destinations & NR_ATTRIBUTE_DESTINATION_LOG)) {
+      nr_attribute_destination_modifier_destroy(&cur);
+      if (NULL != prev) {
+        prev->next = next;
+      } else {
+        config->modifier_list = next;
+      }
+    } else {
+      prev = cur;
+    }
+    cur = next;
+  }
+
+  /* unlikely but if all rules were finalize rules then no more work to do */
+  if (NULL == config->modifier_list) {
+    return;
+  }
+
+  /* now look for any include rules with a destination of
+   * NR_ATTRIBUTE_DESTINATION_LOG and evaluate if any
+   * finalize rules need to be added.
+   */
+  for (cur = config->modifier_list; cur; cur = cur->next) {
+    if (cur->include_destinations & NR_ATTRIBUTE_DESTINATION_LOG) {
+      if (cur->has_wildcard_suffix) {
+        if (0 == cur->match_len) {
+          /* there is an include rule of "*" so no finalize is needed
+           * since all attributes are being explicitely included */
+          return;
+        }
+      }
+      add_finalize_rule = true;
```

Comment:
It looks like you could potentially remove a nesting level here by combining these two if statements:
```suggestion
      if (cur->has_wildcard_suffix && 0 == cur->match_len) {
          /* there is an include rule of "*" so no finalize is needed
           * since all attributes are being explicitely included */
          return;
        }
      add_finalize_rule = true;
```